pkg/agent/discovery_agent.go

```diff
@@ -0,0 +1,255 @@
+	err := c.spokeClient.Get(ctx, req.NamespacedName, hc)
+
+	hcDeleted := false
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.log.Info(fmt.Sprintf("hosted cluster %s/%s is deleted.", req.NamespacedName, req.Name))
+			hcDeleted = true
+		} else {
+			c.log.Error(err, fmt.Sprintf("failed to get the hostedcluster (%s)",
+				req.NamespacedName))
+			return ctrl.Result{}, err
+		}
+	}
+
+	if hcDeleted {
+		c.log.Info(fmt.Sprintf("deleting the discovered cluster for hosted cluster %s", req.NamespacedName))
+		if err := c.deleteDiscoveredCluster(req.Name, req.Namespace, ctx); err != nil {
+			c.log.Error(err, fmt.Sprintf("could not delete discovered cluster for hosted cluster (%s)", hc.Name))
+			return ctrl.Result{}, err
+		}
+	} else {
+		// check if controlplane is available, if not then requeue until it is
+		if hc.Status.Conditions == nil || len(hc.Status.Conditions) == 0 || !isHostedControlPlaneAvailable(hc.Status) {
+			// wait for HCP API server to become available
+			c.log.Info(fmt.Sprintf("hosted control plane of (%s) is unavailable", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+
+		c.log.Info(fmt.Sprintf("creating or updating a discovered cluster for hosted cluster %s", req.NamespacedName))
+
+		if err := c.createUpdateDiscoveredCluster(*hc, ctx); err != nil {
+			c.log.Error(err, fmt.Sprintf("could not create or update discovered cluster for hosted cluster (%s)", hc.Name))
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// creates discovered cluster in the hub cluster
+func (c *DiscoveryAgent) createUpdateDiscoveredCluster(hc hyperv1beta1.HostedCluster, ctx context.Context) error {
+	dcList, err := c.getDiscoveredClusterList(hc, ctx)
+	if err != nil {
+		return err
+	}
+
+	newDc := c.getDiscoveredCluster(hc)
+
+	if len(dcList.Items) == 0 {
+		c.log.Info(fmt.Sprintf("creating discovered cluster for hosted cluster (%s/%s)", hc.Namespace, hc.Name))
+
+		if err = c.hubClient.Create(ctx, newDc, &client.CreateOptions{}); err != nil {
+			c.log.Error(err, fmt.Sprintf("failed to create discovered cluster for hosted cluster (%s/%s)", hc.Namespace, hc.Name))
+			return err
+		}
+	} else if len(dcList.Items) == 1 {
+		oldDc := dcList.Items[0]
+
+		// As a hosted cluster gets updated by the hypershift operator, the API URL and the Openshift version
+		// are the only ones in DiscoveredCluster spec that could be changed.
+		if oldDc.Spec.APIURL != newDc.Spec.APIURL || oldDc.Spec.OpenshiftVersion != newDc.Spec.OpenshiftVersion {
+			c.log.Info(fmt.Sprintf("updating discovered cluster for hosted cluster (%s/%s)", hc.Namespace, hc.Name))
+
+			dc := oldDc.DeepCopy()
+			dc.Spec.APIURL = newDc.Spec.APIURL
+			dc.Spec.OpenshiftVersion = newDc.Spec.OpenshiftVersion
+			if err = c.hubClient.Update(ctx, dc); err != nil {
+				c.log.Error(err, fmt.Sprintf("failed to update discovered cluster for hosted cluster (%s/%s)", hc.Namespace, hc.Name))
+				return err
+			}
+		}
+	} else {
```

Comment:
Would you add comment for this logic, why not having discoveredClusters previously and a new discoveredCluster added as well ?